tests/common/util.rs

```diff
@@ -1359,26 +1351,93 @@ impl UCommand {
         self
     }
 
-    /// Spawns the command, feeds the stdin if any, and returns the
-    /// child process immediately.
-    pub fn run_no_wait(&mut self) -> UChild {
-        assert!(!self.has_run, "{}", ALREADY_RUN);
-        self.has_run = true;
-        log_info("run", &self.comm_string);
+    fn build(&mut self) -> (Command, Option<CapturedOutput>, Option<CapturedOutput>) {
+        if self.bin_path.is_some() {
+            if let Some(util_name) = &self.util_name {
+                self.args.push_front(util_name.into());
+            }
+        } else if let Some(util_name) = &self.util_name {
+            self.bin_path = Some(PathBuf::from(TESTS_BINARY));
+            self.args.push_front(util_name.into());
+        // neither `bin_path` nor `util_name` was set so we apply the default to run the arguments
+        // in a shell platform specific
```

Comment:
This comment seems to be unfinished? Also maybe we want to panic if neither was set? Also, why would the `bin_path` not be set? Can't we just make it a `PathBuf` and force it to be set in any constructor?

Reply:
This comment is actually finished. It's just a helping comment to guide the quick reader a little bit through the build method.

> Also maybe we want to panic if neither was set? I'm also wondering why the bin_path would not be set? Can't we just make it a PathBuf and force it to be set in any constructor?

The `build` method is built on the concept to collect all the information given by the user up to this call of `build` and then apply the defaults appropriately. So, `bin_path` is always set (a few lines later, you'll find another helping comment above the `unwrap` which points this out). The application of defaults only in `build` simplifies the building process and keeps `UCommand` as flexible as possible (also for future changes). I think it's much clearer to apply the defaults in a single, dedicated method instead of spreading the setting of defaults over the `init` methods and the `build` method.

I'll place a doc comment above the `build` method which will describe this strategy in detail.

Reply:
ok done.

Reply:
Does this answer your questsions?

Reply:
Ah I understand the comment now. It was the word order that tripped me up. "in a platform specific shell" might be clearer. I understand the builder, but I just think the command name should always be given in the constructor of the builder. If we do that the bin_name does not need to be an option anymore.

Reply:
> ...  but I just think the command name should always be given in the constructor of the builder ...

Why? What's the harm with `bin_path` being an `Option` and applying all the defaults in `build`? I don't think we should or need to replicate `std::process::Command`. I actually think, we should instead do it our own way to better match our requirements. What if we need `bin_path()` public in a year or two, or introduce other builder methods? Such changes can be easily implemented and we have a clear and tidy design.

Reply:
> Why?

We're already using it that way with the `cmd`, `ccmd`, `ucmd` functions and it simplifies the logic a bit.

> we should instead do it our own way to better match our requirements

How does it match better?

Reply:
> We're already using it that way with the cmd, ccmd, ucmd functions and it simplifies the logic a bit.

In `TestScenario` yes, but that's a different struct and isn't a builder. However, I thought about simplifying the logic in `build`. Is this your main concern? Would splitting the `build` method into logically related smaller methods help, too?

> How does it match better?

In short, we have a good default for a non-existent `bin_path`, `Command` doesn't. I actually already said that, but I also think it wouldn't harm to keep `UCommand` clear and open for changes.

Reply:
> Would splitting the build method into logically related smaller methods help, too?

No the build method is fine!

> In short, we have a good default for a non-existent bin_path, Command doesn't.

Well, that's the thing, I don't want there to be a default, because I don't see the point of doing this. It would be like calling a function but not specifying the function name but only the arguments, wouldn't it?

Reply:
That being said, if you still disagree it's not a big deal and we can move on.

Reply:
Let me try to give an example. If I understand this PR correctly, this would be valid:
```rust
UCommand::new()
    .arg("foo")
    .arg("bar");
```
Under the hood, this would become `sh -c foo bar`, so that's the same as calling `foo bar` which the shell option. That's very weird to me and I'd much rather have this be an error (and preferably at compile time). Also remember that flexibility is not really the point here. We're building these tools for a very specific purpose, so we get to tailor it to our (current) preferences. If the requirements change, this will just change with it.

Reply:
yeah the code above would be valid. I actually would prefer providing a default if possible over errors. Would renaming `new` to `with_shell` (and giving up the `new` method) resolve your concerns?

Reply:
Not really, but I just realized there's some additional complexity because there's both `bin_name` and `util_name` and I need to think for a bit about how we want those to interact with the shell. There are some fundamental problems with doing this platform-independent shell API and how it interacts with the tests binary.